RecoPixelVertexing/PixelTrackFitting/plugins/PixelTrackReconstructionGPU.cu

```diff
@@ -1,124 +1,210 @@
+                     Rfit::line_fit *line_fit) {
+
+  // Reshape Eigen components from hits_and_covariances, using proper thread and
+  // block indices Perform the fit Store the results in the proper vector, using
+  // again correct indices
+
+  // Loop for hits_in_fit times:
+  //   first 3 are the points
+  //   the rest is the covariance matrix, 3x3
+  int start = (blockIdx.x * blockDim.x + threadIdx.x) * hits_in_fit * 12;
+  int helix_start = (blockIdx.x * blockDim.x + threadIdx.x);
+  if (start >= cumulative_size) {
+    return;
+  }
+
+  if (GPU_DEBUG) {
+    printf("BlockDim.x: %d, BlockIdx.x: %d, threadIdx.x: %d, start: %d, "
+           "cumulative_size: %d\n",
+           blockDim.x, blockIdx.x, threadIdx.x, start, cumulative_size);
+  }
+
+  line_fit[helix_start] =
+      Rfit::Line_fit(hits[helix_start], hits_cov[helix_start],
+                     circle_fit[helix_start], fast_fit[helix_start], true);
+
+  par_uvrtopak(circle_fit[helix_start], B, true);
 
   // Grab helix_fit from the proper location in the output vector
   Rfit::helix_fit &helix = results[helix_start];
-  helix.par << circle.par, line.par;
+  helix.par << circle_fit[helix_start].par, line_fit[helix_start].par;
   // TODO: pass properly error booleans
-  if (1) {
+
     helix.cov = MatrixXd::Zero(5, 5);
-    helix.cov.block(0, 0, 3, 3) = circle.cov;
-    helix.cov.block(3, 3, 2, 2) = line.cov;
+    helix.cov.block(0, 0, 3, 3) = circle_fit[helix_start].cov;
+    helix.cov.block(3, 3, 2, 2) = line_fit[helix_start].cov;
+
+  helix.q = circle_fit[helix_start].q;
+  helix.chi2_circle = circle_fit[helix_start].chi2;
+  helix.chi2_line = line_fit[helix_start].chi2;
+
+  if (GPU_DEBUG) {
+    printf("KernelLineFitAllHits line.par(0): %d %f\n", helix_start,
+           circle_fit[helix_start].par(0));
+    printf("KernelLineFitAllHits line.par(1): %d %f\n", helix_start,
+           line_fit[helix_start].par(1));
   }
-  helix.q = circle.q;
-  helix.chi2_circle = circle.chi2;
-  helix.chi2_line = line.chi2;
 }
 
-void PixelTrackReconstructionGPU::launchKernelFit(float * hits_and_covariancesGPU, 
-    int cumulative_size, int hits_in_fit, float B, Rfit::helix_fit * results) {
-  const dim3 threads_per_block(32,1);
-  // We need to partition data in blocks of:
-  // 12(3+9) * hits_in_fit
-  int num_blocks = cumulative_size/(hits_in_fit*12)/threads_per_block.x + 1;
-  KernelFullFitAllHits<<<num_blocks, threads_per_block>>>(hits_and_covariancesGPU, hits_in_fit, cumulative_size, B, results);
+void PixelTrackReconstructionGPU::launchKernelFit(
+    float *hits_and_covariancesGPU, int cumulative_size, int hits_in_fit,
+    float B, Rfit::helix_fit *results) {
+  const dim3 threads_per_block(32, 1);
+  int num_blocks =
+      cumulative_size / (hits_in_fit * 12) / threads_per_block.x + 1;
+  auto numberOfSeeds = cumulative_size / (hits_in_fit * 12);
+  Rfit::Matrix3xNd *hitsGPU;
+  cudaCheck(cudaMalloc((void **)&hitsGPU,
```

Comment:
Do we need to call `cudaMalloc(...)` at every event ?
Can we preallocate the buffers as in the other modules ?

Reply:
Maybe it is not worth of reorganizing as the plan is to copy the calls to the fit kernels to the CA EDProducer (and in there preallocate the memory).

Reply:
exactly, I left things as they were before. Moving everything to the CA will also include the memory preallocation.